azure-core/azure/core/pipeline/__init__.py

```diff
@@ -47,10 +47,11 @@ def __exit__(self, exc_type, exc_value, traceback):
 
 class PipelineContext(dict):
 
-    def __init__(self, transport, **kwargs):
+    def __init__(self, session, transport, **kwargs):
```

Comment:
I cannot find out the reason that we need to expose session to outside of the transport

Reply:
Because if multiple threads are using the pipeline, and one finishes first, that session will be closed while another is still using it.
If the session is external to the transport then we have one thread = one session and that session can be opened and closed without impacting others.

Reply:
Do you mean one transport object may have multiple session members?

Reply:
The transport would probably need to manage multiple sessions yeah. Whereas we manage the session outside of the transport then we can just associate a session with each pipeline.run

Reply:
In which case one transport object needs to manage multiple sessions? Why not create multiple transport objects?

Reply:
With the assumption of "base url never changes after the first set", I think we can implement a smart session:
ctor:
self.session = None
self.refCount = 0
enter:
self.refCount ++
if self.session is None
self.session = new session()

exit:
self.refCount --
if self.refCount == 0
self.session.close
self.session = None